### DISCUSSION TITLE
How to integrate into an actual DNS server?

### DISCUSSION THREAD

**Discussion by davehouser1** [sys4/netbox-plugin-dns#19] [Q&A] [answered]
Does this plugin have capabilities to integrate into an active DNS server to manage it? If so what are the steps to do so? Main Microsoft AD?

I assume this plugin is just an overlay, but nothing in the documentation I found actually details it. Can you share more details?
What would be really nice is if Netbox ran its own DNS server and I could manage it from here. Not sure if possible?

**Comment by peteeckel** [comment]
Hi @davehouser1, thanks for asking. I've moved this issue to a discussion because the question arises sometimes, and has been answered a couple of times ago. I hope the answer is easier to find here than in an issue that sooner or later gets closed and disappears from view.

Generally, NetBox is widely understood as a "Source of Truth" for Network and Data Center information, not actually a management tool to manage aspects of this (e.g. it's not a management tool for switches and routers, neither it is meant to be used to manage your virtualisation environment, start VMs, assign memory and disks and so on. 

In just the same way, NetBox DNS is not meant to be a DNS server UI, much less a DNS server itself. The main purpose of NetBox DNS is to maintain DNS data, automatically create NS and SOA records, maintain PTR records for address records, validate DNS records and so on, just like NetBox is used to manage IP addresses, rack space interfaces and cables. It's all about the data, not the actual implementation - the data in NetBox DNS can be used with every DNS server implementation in every possible infrastructure (at least that what it's striving for, input where it doesn't work is always welcome!).

That said, there are many ways to make use of the DNS data maintained in NetBox DNS. You can export zone files using Custom Scripts (quite easily, actually), and when you're running a local DNS server on your NetBox machine that's already pretty close to what you wish for. Look at the `NetBox_DNS_Exporter`example script for a simple blueprint how to do that.

For a larger DNS infrastructure you'll also want to configure slaves, resolvers etc. I wouldn't try to use Custom Scripts for that (though you definitely can - they are really universal in what they can do, being limited only by the capabilities of Python) but leave it to a configuration management tool such as Ansible (there is an Ansible example in the repository as well, and some documentation). If that's not general enough, there are still the REST and GraphQL APIs that give unlimited access to the data in NetBox (and, as a plugin, NetBox DNS as well).

The main intention behind not supporting any particular DNS implementation is flexibility: NetBox DNS is agnostic to the DNS server you are using, so you can use any DNS server you want (and, that's the downside, see how you integrate both). 

I'll be glad to suggest viable solutions (more examples never hurt!) if you have a specific use case you can't figure out how to cover.

**Reply by davehouser1** [reply]
@peteeckel thank you for quick response. Our environment and the specific configuration solution I am looking to set up is the following:

- We have MS Windows Server Active Directory DC's in our env. They are the primary and secondary DNS servers all our hosts to use. 
- We want to use an IPAM that has the ability create DNS records (like your plugin does).
- We want most** DNS records to be created form the IPAM. To be more specific, in Netbox, from the home page IPAM > IP Address > Add new > configure DNS Name (with FQDN)***. We want this to create an A and PTR record in specific MS Windows Server DNS Zones. 
- Do not want to have to set up secondary sub domains / zones if not necessary. would like to use the same domain our DC's use. 

We want to use an IPAM to be a source of all our data. But right now we would need to manage DNS in two locations (Directly on our DCs and the IPAM) which is getting annoying. 

I have worked with Bind in the past, and know how to stand up a working server. I also have extensive knowledge working with Ansible, however Netbox Swagger UI shows so many ways I can interface with the systems API too (which is great!) so not sure I need to use Ansible to configure Netbox directly. I know what I want is possible, albite will take some work to set up. Someone suggested I set up my own DNS server on the same host that Netbox runs on (maybe Bind?). Then set up zone transfers between Netbox/bind and windows DC's. If I can get that to work, then its figuring out how to automate Netbox to do the following:

1. pull in DNS records from the zone transfer
2. send local DNS records from the zone transfer
3. get Netbox to actual build records from the IPAM > IP Address > DNS Name

I really want DNS to do all the replication work. I don't want to have to set up WINRM or SSH on the DC's if I don't need too. 

What do you think? Is this all possible? Any recommendations would be very helpful to start me out on my journey :)

** We do integrate hosts to Active Directory (Linux and Windows), when doing so, windows automatically creates DNS records for each. So these systems would be created indirectly through AD integration process. I would be really great if we could create DNS records on our DC's or the IPAM and they replicate between all of them.

 *** I set up a nameserver and a zone in the DNS plugin location. I was also able to build a A record under records for the plugin. However I noticed that if I try to populate IPAM > IP Address > DNS Name, the IP Address gets built, but no records show up under Netbox DNS records, is this expected?

**Comment by ThomasADavis** [comment]
We use this -

https://github.com/jcollie/octodns-netbox-dns

with octodns - https://github.com/octodns/octodns

This allows a sync between netbox dns plugin and dns servers.  They don't seem to have a plugin to talk to windows dns servers.. But if your capable of generating bind files, octodns can pickup the bind files and populate netbox dns plugin entries for you.

**Reply by peteeckel** [reply]
My understanding is that @davehouser1 has the problem of getting data into Azure AD, not out of it - otherwise the zone files can easily be generated from AXFR requests against Azure AD (as with any other DNS server).

It would, however, be very interesting to know if there is an OctoDNS plugin for Azure AD around (or if there is reasonably simple way to create one). That would, together with the OctoDNS/NetBox DNS plugin, indeed be a good solution for the problem at hand IMHO.

**Comment by peteeckel** [comment, marked as answer]
Hi @davehouser1, I'm not experienced with MS Azure AD, but if it fits your use case I would work along the following architecture: 

1. Records are maintained in NetBox DNS, PTR created automatically
2. There is a local authoritative DNS server on the NetBox machine or any other machine with a shared file system
3. Either via Webhook (on zone update, probably) or manually, zone files are created and a reload is triggered on the local DNS server
4. Azure AD gets the zones via NOTIFY/AXFR

This is based on a couple of assumptions: First, zone creations or deletions are assumed to be a rare occasion. For creating or deleting a zone, the config files for BIND and some magic in Azure AD needs to be performed. If you don't want to do it manually (which is perfectly OK if it happens only a handful of times per year) you'll need some kind of config management tool in addition.

Then there is the assumption that record updates don't happen extremely frequently. The problem with Webhooks is that they are triggered synchronously, i.e. directly on, for instance, a zone update event. Updating an A record will result in another update to the associated PTR record, and two updates to the forward and reverse zone for the SOA SERIAL field. While dumping the zone files and sending a reload command to the name server four times in a row is ugly, but not a major issue, it would quickly get a problem when you insert tens or hundreds of records on a regular base. In this case you'll need a two-stage approach: One stage just creates some kind or semaphore, like a flag file or timer, the second stage runs asynchronously based on a time schedule and updates the name server if the flag file is present or the timer runs out. This could be done via cron (`manage.py runscript ...`) or via the `netbox-rq` service, i.e. scheduled script.

Another way to solve the latter issue would be to run the export manually via Custom Script, but that has the disadvantage of depending on the user remembering to do it - so some kind of monitoring would be needed.

You can also have a look at the example script `DNS_IPAM_Updater.py` for an idea how to synchronise the IPAM 'DNS Name' field from NetBox DNS. Using IPAM as a source for DNS is not too powerful as you only can create address records this way, not CNAME/TXT/NS etc., so except when this doesn't matter and you only need to manage name/address records I would not go down that way (about two years ago that was my original idea for a customer's DNS management, but I quickly disregarded that idea and found `netbox-dns` instead).

**Comment by bitcollector1** [comment]
Thank you so much @peteeckel for all the hard work you and others have put into this amazing plugin!  

I used the example script (Zone Exporter) and modified it a bit to send the zone files to my docker instance of BIND9 and then restart the docker service. It's all working so much better than I had hoped it would; I learned a lot about DNS in the process. 

This plugin, along with the kea one has really changed the game for managing dynamic resources in a fast moving lab. 

https://github.com/devon-mar/netbox-kea

**Reply by bitcollector1** [reply]
I have some ideas for extending the functionality of netbox-kea plugin above. We have premium hooks and can do some cool things with that. Hopefully I can get some time this year to build my first plugin.
